### PR TITLE
survey transcriptome index strain mapping on assembly

### PR DESCRIPTION
**Pull request by davidsmejia**
## Issue Number

#3381 

## Purpose/Implementation Notes

- took logic that fetches and searches results into a static method on transcriptome index surveyor
- change logic when surveying fungi/bacteria from checking name to checking assembly
- override the `species_taxonomy_id` for bacteria and fungi
- throw an error if more than one assembly matches when surveying transcriptome (requires manual intervention)
- merge and extend tests to check if surveyed files exist remotely
- fix issue for bacteria download files where assembly has an underscore after in file name
- recreate the cached requests cassettes

## Methods

In order to create the correct Bacteria and Fungi this PR contains a change that requires us to include the `species_taxonom_id` in `config/organism_strain_mapping.csv`. This value should be the NCBI taxonomy ID of the organism that we want the strain to represent. This value comes back matching for bacteria so we must specify it now or we will create strain organisms when creating / saving to the DB.

## Types of changes

- New feature (non-breaking change which adds functionality)
- Bug Fix

## Functional tests

n/a

## Checklist

- [x] Lint and unit tests pass locally with my changes

## Screenshots

n/a

### CODE REVIEW

**Comment by davidsmejia** [comment]
Downloader and No-op seem to be failing due to missing `ascp` / being unable to install. I think this is good to review at this time.

File: foreman/data_refinery_foreman/surveyor/transcriptome_index.py [hunk 1]
```diff
@@ -403,51 +447,15 @@ def discover_species(self):
...
-        if len(all_new_species) == 0:
+        if len(discovered_species) == 0:
             logger.error(
                 "Unable to find any species!",
                 ensembl_division=ensembl_division,
                 organism_name=organism_name,
             )
 
-        return all_new_species
+        return list(map(self._generate_files, discovered_species))
```
Comment (arkid15r) [inline foreman/data_refinery_foreman/surveyor/transcriptome_index.py:461]: The `def _generate_files(self, species: Dict) -> None:` is out of date

File: foreman/data_refinery_foreman/surveyor/utils.py [hunk 2]
```diff
@@ -1,11 +1,40 @@
...
+    """
+    for item in iterable:
+        if item.get(key) == value:
+            return item
+
+
+def filter_dicts_on_key(key: str, value: Any, iterable: Iterable) -> Generator:
+    """
+    Takes a key, value, and iterable of dictionaries and returns
+    a Generator that yields dictionaries that matched on key and value.
```
Comment (arkid15r) [inline foreman/data_refinery_foreman/surveyor/utils.py:31]:
```suggestion
    Takes a key, value, and iterable of dictionaries and yields
    dictionaries that matched on key and value.
```

File: foreman/tests/surveyor/test_utils.py [hunk 3]
```diff
@@ -29,3 +29,64 @@ def test_flatten_dict(self):
...
+            {"test_key": 4},
+        ]
+
+        # Test for successful match.
+        matched_dict = utils.find_first_dict("test_key", 1, iterable_of_dicts)
+        self.assertEqual(matched_dict["test_key"], 1)
+
+        # Test when key exists, but no match.
+        good_key_no_match = utils.find_first_dict("test_key", 5, iterable_of_dicts)
+        self.assertEqual(good_key_no_match, None)
```
Comment (arkid15r) [inline foreman/tests/surveyor/test_utils.py:51]:
```suggestion
        self.assertIsNone(good_key_no_match)
```

File: foreman/tests/surveyor/test_utils.py [hunk 4]
```diff
@@ -29,3 +29,64 @@ def test_flatten_dict(self):
...
+
+        # Test when key does not exist.
+        no_dict = utils.find_first_dict("another_key", 1, iterable_of_dicts)
+        self.assertEqual(no_dict, None)
```
Comment (arkid15r) [inline foreman/tests/surveyor/test_utils.py:55]:
```suggestion
        self.assertIsNone(no_dict)
```

File: foreman/tests/surveyor/test_utils.py [hunk 5]
```diff
@@ -29,3 +29,64 @@ def test_flatten_dict(self):
...
+
+        # Test when iterable is empty.
+        empty_iterable_dict = utils.find_first_dict("test_key", 1, [])
+        self.assertEqual(empty_iterable_dict, None)
```
Comment (arkid15r) [inline foreman/tests/surveyor/test_utils.py:59]:
```suggestion
        self.assertIsNone(empty_iterable_dict)
```

**Review by arkid15r — APPROVED** [review]
LGTM 👍

**Pull request merged**